x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/query/LongScriptFieldDistanceFeatureQuery.java

```diff
@@ -0,0 +1,211 @@
+import org.elasticsearch.script.Script;
+import org.elasticsearch.xpack.runtimefields.AbstractLongScriptFieldScript;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Set;
+
+public final class LongScriptFieldDistanceFeatureQuery extends AbstractScriptFieldQuery {
+    private final CheckedFunction<LeafReaderContext, AbstractLongScriptFieldScript, IOException> leafFactory;
+    private final long origin;
+    private final long pivot;
+    private final float boost;
+
+    public LongScriptFieldDistanceFeatureQuery(
+        Script script,
+        CheckedFunction<LeafReaderContext, AbstractLongScriptFieldScript, IOException> leafFactory,
+        String fieldName,
+        long origin,
+        long pivot,
+        float boost
+    ) {
+        super(script, fieldName);
+        this.leafFactory = leafFactory;
+        this.origin = origin;
+        this.pivot = pivot;
+        this.boost = boost;
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        return new Weight(this) {
+            @Override
+            public boolean isCacheable(LeafReaderContext ctx) {
+                return false;
+            }
+
+            @Override
+            public void extractTerms(Set<Term> terms) {}
+
+            @Override
+            public Scorer scorer(LeafReaderContext context) throws IOException {
+                return new DistanceScorer(this, leafFactory.apply(context), context.reader().maxDoc(), boost);
+            }
+
+            @Override
+            public Explanation explain(LeafReaderContext context, int doc) throws IOException {
+                AbstractLongScriptFieldScript script = leafFactory.apply(context);
+                script.runForDoc(doc);
+                long value = valueWithMinAbsoluteDistance(script);
+                float weight = LongScriptFieldDistanceFeatureQuery.this.boost * boost;
+                float score = score(weight, distanceFor(value));
+                return Explanation.match(
+                    score,
+                    "Distance score, computed as weight * pivot / (pivot + abs(value - origin)) from:",
+                    Explanation.match(weight, "weight"),
+                    Explanation.match(pivot, "pivot"),
+                    Explanation.match(origin, "origin"),
+                    Explanation.match(value, "current value")
+                );
+            }
+        };
+    }
+
+    private class DistanceScorer extends Scorer {
+        private final AbstractLongScriptFieldScript script;
+        private final TwoPhaseIterator twoPhase;
+        private final DocIdSetIterator disi;
+        private final float weight;
+
+        protected DistanceScorer(Weight weight, AbstractLongScriptFieldScript script, int maxDoc, float boost) {
+            super(weight);
+            this.script = script;
+            twoPhase = new TwoPhaseIterator(DocIdSetIterator.all(maxDoc)) {
```

Comment:
I am not familiar with runtime fields, but I am wondering if we intend always to create an iterator across all documents? Do we plan to add support to limit number of docs (e.g. only docs returned by a top filter)?

Reply:
I believe we're hoping for `bool` queries to AND together a "normal" query and a runtime field query. I've experimented with this for our `term` and `match` style queries and it seems to work pretty well. If the "normal" query is selective then the runtime query won't be asked if it matches most documents. On the flip side, if the runtime field query non-selective then we'll quickly fill up the 10,000 hits and terminate early.